File: services/web/src/types/camera.ts
export interface Camera {
  id: string;
  name: string;
  ip_address: string;
  mac_address: string | null;
  manufacturer: string | null;
  model: string | null;
  firmware_version: string | null;
  serial_number: string | null;
  stream_main_uri: string | null;
  stream_sub_uri: string | null;
  stream_audio_uri: string | null;
  auth_type: string;
  username: string;
  has_audio: boolean;
  has_talkback: boolean;
  has_ptz: boolean;
  has_onvif: boolean;
  has_motion_detection: boolean;
  has_io_ports: boolean;
  motion_source: string | null;
  max_resolution: string | null;
  recording_mode: string;
  stream_transport: string;
  ptz_presets: string[] | null;
  status: string;
  connection_error: string | null;
  last_seen_at: string | null;
  tags: string[] | null;
  location: string | null;
  location_id: string | null;
  location_name: string | null;
  storage_backend_id: string | null;
  storage_backend_name: string | null;
  notes: string | null;
  privacy_mode: string | null;
  ai_enabled: boolean;
  ai_objects: string[] | null;
  ai_zones: { points: [number, number][] }[] | null;
  ai_sensitivity: string;
  ai_min_confidence: number;
  display_order: number;
  created_at: string;
  updated_at: string;
}

export interface CameraReorderItem {
  id: string;
  display_order: number;
}

export interface Location {
  id: string;
  name: string;
  description: string | null;
  color: string;
  camera_count: number;
  created_at: string;
}

export interface CameraCreatePayload {
  name: string;
  ip_address: string;
  username: string;
  password?: string;
  auth_type: string;
  stream_main_uri?: string;
  stream_sub_uri?: string;
  stream_audio_uri?: string;
  recording_mode: string;
  stream_transport: string;
  tags?: string[];
  location?: string;
  location_id?: string | null;
  storage_backend_id?: string;
  notes?: string;
  ai_enabled?: boolean;
  ai_objects?: string[];
  ai_sensitivity?: string;
  ai_min_confidence?: number;
}

export interface CameraUpdatePayload {
  name?: string;
  ip_address?: string;
  username?: string;
  password?: string;
  auth_type?: string;
  stream_main_uri?: string;
  stream_sub_uri?: string;
  stream_audio_uri?: string;
  recording_mode?: string;
  stream_transport?: string;
  is_active?: boolean;
  tags?: string[];
  location?: string;
  location_id?: string | null;
  storage_backend_id?: string;
  notes?: string;
  privacy_mode?: string;
  ai_enabled?: boolean;
  ai_objects?: string[];
  ai_zones?: { points: [number, number][] }[] | null;
  ai_sensitivity?: string;
  ai_min_confidence?: number;
  motion_source?: string;
}

export interface ProbeResult {
  reachable: boolean;
  ip: string;
  open_ports: number[];
  manufacturer: string | null;
  model: string | null;
  server_header: string | null;
  http_title: string | null;
  stream_main_uri: string | null;
  stream_sub_uri: string | null;
  has_rtsp: boolean;
  has_http: boolean;
  has_audio: boolean;
  has_ptz: boolean;
  has_onvif: boolean;
  has_motion_detection: boolean;
  onvif_url: string | null;
}

export interface DiscoveredDevice {
  ip_address: string;
  manufacturer: string | null;
  model: string | null;
  http_title: string | null;
  stream_main_uri: string | null;
  open_ports: number[];
  has_rtsp: boolean;
  has_http: boolean;
  has_audio: boolean;
  has_ptz: boolean;
  has_onvif: boolean;
  has_motion_detection: boolean;
  confidence: number;
}

export interface DiscoveryStatus {
  scan_id: string;
  status: string;
  phases: Record<string, string>;
  found_count: number;
  progress_pct: number;
  scanned_ips: number;
  total_ips: number;
}

export interface TestResult {
  reachable: boolean;
  rtsp_ok: boolean;
  auth_ok: boolean;
  error_code: string | null;
  error_message: string | null;
  latency_ms: number | null;
  stream_resolution: string | null;
  stream_codec: string | null;
  manufacturer: string | null;
  open_ports: number[];
}

File: services/api/app/schemas/camera.py
"""Camera Pydantic schemas."""

from pydantic import BaseModel, Field


class CameraCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    ip_address: str
    username: str = "admin"
    password: str | None = None
    auth_type: str = "basic"
    stream_main_uri: str | None = None
    stream_sub_uri: str | None = None
    stream_audio_uri: str | None = None
    recording_mode: str = "continuous"
    stream_transport: str = "tcp"
    tags: list[str] | None = None
    location: str | None = None
    location_id: str | None = None
    notes: str | None = None


class CameraUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    ip_address: str | None = None
    username: str | None = None
    password: str | None = None
    auth_type: str | None = None
    stream_main_uri: str | None = None
    stream_sub_uri: str | None = None
    stream_audio_uri: str | None = None
    recording_mode: str | None = None
    stream_transport: str | None = None
    is_active: bool | None = None
    tags: list[str] | None = None
    location: str | None = None
    location_id: str | None = None
    notes: str | None = None
    privacy_mode: str | None = None


class CameraResponse(BaseModel):
    id: str
    name: str
    ip_address: str
    mac_address: str | None = None
    manufacturer: str | None = None
    model: str | None = None
    firmware_version: str | None = None
    serial_number: str | None = None
    stream_main_uri: str | None = None
    stream_sub_uri: str | None = None
    stream_audio_uri: str | None = None
    auth_type: str
    username: str
    has_audio: bool
    has_talkback: bool
    has_ptz: bool
    has_onvif: bool
    has_motion_detection: bool
    has_io_ports: bool
    motion_source: str | None = None
    max_resolution: str | None = None
    recording_mode: str
    stream_transport: str
    ptz_presets: list | None = None
    status: str
    connection_error: str | None = None
    last_seen_at: str | None = None
    tags: list[str] | None = None
    location: str | None = None
    location_id: str | None = None
    location_name: str | None = None
    notes: str | None = None
    privacy_mode: str | None = None
    created_at: str
    updated_at: str


class DiscoveryRequest(BaseModel):
    subnets: list[str] = ["192.168.1.0/24"]
    methods: list[str] | None = None
    timeout: int = Field(default=120, ge=10, le=600)


class ProbeRequest(BaseModel):
    ip_address: str


class ProbeResponse(BaseModel):
    reachable: bool
    ip: str
    open_ports: list[int]
    manufacturer: str | None = None
    model: str | None = None
    server_header: str | None = None
    http_title: str | None = None
    stream_main_uri: str | None = None
    has_rtsp: bool = False
    has_http: bool = False
    has_audio: bool = False
    has_ptz: bool = False
    has_onvif: bool = False
    has_motion_detection: bool = False


class DiscoveryStatusResponse(BaseModel):
    scan_id: str
    status: str
    phases: dict
    found_count: int
    progress_pct: int
    scanned_ips: int = 0
    total_ips: int = 0


class DiscoveredDevice(BaseModel):
    ip_address: str
    manufacturer: str | None = None
    model: str | None = None
    http_title: str | None = None
    stream_main_uri: str | None = None
    open_ports: list[int] = []
    has_rtsp: bool = False
    has_http: bool = False
    has_audio: bool = False
    has_ptz: bool = False
    has_onvif: bool = False
    has_motion_detection: bool = False
    confidence: int = 0


class CameraTestResponse(BaseModel):
    reachable: bool
    rtsp_ok: bool
    auth_ok: bool = False
    error_code: str | None = None
    error_message: str | None = None
    latency_ms: int | None = None
    stream_resolution: str | None = None
    stream_codec: str | None = None
    manufacturer: str | None = None
    open_ports: list[int] = []
